doc/source/docs/user_guide/spatial_indexing.rst

```diff
@@ -0,0 +1,250 @@
+    points.plot(ax=axs[0], color="limegreen", zorder=3, edgecolor="black", linewidth=.5)
+    points.plot(ax=axs[1], color="limegreen", zorder=3, edgecolor="black", linewidth=.5)
+    nyc.iloc[pred_inside].plot(ax=axs[0], color='orange')
+    @savefig predicate.png
+    nyc.envelope.iloc[pred_inside].plot(ax=axs[1], color='orange');
+
+
+You can use any of the predicates available in :attr:`~sindex.SpatialIndex.valid_query_predicates`:
+
+.. ipython:: python
+
+    nyc.sindex.valid_query_predicates
+
+Array query
+~~~~~~~~~~~
+
+Checking a single geometry against the tree is nice but not that
+efficient if you are interested in many-to-many relationships. The
+:meth:`~sindex.SpatialIndex.query` method allows passing any 1-D array of geometries to be
+checked against the tree. If you do so, the output structure is slightly
+different:
+
+.. ipython:: python
+
+    bbox_array_query = nyc.sindex.query(points)
+    bbox_array_query
+
+
+By default, the method returns a 2-D array of indices where the query
+found a hit where the subarrays correspond to the indices of the input
+geometries and indices of the tree geometries associated with each. In
+the example above, the 0-th geometry in the ``points`` GeoSeries
+intersects the bounding box of the geometry at the position 1 from the
+``nyc`` GeoDataFrame, while the geometry 1 in the ``points`` matches
+geometry 16 in the ``nyc``. You may notice that these are the same
+indices as you’ve seen above.
+
+The other option is to return a boolean array with shape
+``(len(tree), n)`` with boolean values marking whether the bounding box
+of a geometry in the tree intersects a bounding box of a given geometry.
+This can be either a dense numpy array, or a sparse scipy array. Keep in
+mind that the output will be, in most cases, mostly filled with
+``False`` and the array can become really large, so it is recommended to
+use the sparse format, if possible.
+
+You can specify each using the ``output_format`` keyword:
+
+.. ipython:: python
+
+    bbox_array_query_dense = nyc.sindex.query(points, output_format="dense")
+    bbox_array_query_dense
+
+
+The dense array above has rows aligned with the rows of ``nyc`` and
+columns aligned with the rows of ``points`` and indicates all pairs
+where a *hit* was found.
+
+The same array can be represented as a :func:`scipy.sparse.coo_array`:
+
+.. ipython:: python
+
+    bbox_array_query_sparse = nyc.sindex.query(points, output_format="sparse")
+    bbox_array_query_sparse
+
+One example illustrating the usage is a task of finding the number of
+neighboring geometries for each subborough. You can use the spatial
+index to check all geometries against each other.
+
+.. ipython:: python
+
+    neighbors = nyc.sindex.query(nyc.geometry, predicate="intersects", output_format="sparse")
+    neighbors
+
+Since you are using ``nyc`` on both sides of the query here, the
+resulting array is square-shaped with diagonal filled with ``True``.
```

Comment:
The rendered output is not square shaped, so it is harder to see the pattern described here.  Is it possible to reshape the output to ensure it renders as a square matrix?

Reply:
This refers to the output of the cell above which is a square, although sparse. I'll try to be more precise about what this refers to and maybe use the dense output as it might be clearer when talking about square shape.